augur/merge.py

```diff
@@ -0,0 +1,303 @@
+
+                f'create unique index {sqlite_quote_id(f"{m.table_name}_id")} on {sqlite_quote_id(m.table_name)}({sqlite_quote_id(m.id_column)});',
+
+                # <https://sqlite.org/pragma.html#pragma_optimize>
+                f'pragma optimize;')
+
+            # We're going to use Metadata.columns to generate the select
+            # statement, so ensure it matches what SQLite's .import created.
+            assert m.columns == (table_columns := sqlite3_table_columns(db_path, m.table_name)), \
+                f"{m.columns!r} == {table_columns!r}"
+
+            # Track which columns appear in which metadata inputs, preserving
+            # the order of both.
+            for column in m.columns:
+                # Match different id column names in different metadata files
+                # since they're logically equivalent.
+                output_column = output_id_column if column == m.id_column else column
+
+                output_columns.setdefault(output_column, [])
+                output_columns[output_column] += [(m.table_name, column)]
+
+
+        # Construct query to produce merged metadata.
+        select_list = [
+            # Output metadata columns coalesced across input metadata columns
+            *(f"""coalesce({', '.join(f"nullif({x}, '')" for x in starmap(sqlite_quote_id, reversed(input_columns)))}, null) as {sqlite_quote_id(output_column)}"""
+                for output_column, input_columns in output_columns.items()),
+
+            # Source columns
+            *(f"""{sqlite_quote_id(m.table_name, m.id_column)} is not null as {sqlite_quote_id(f'__source_metadata_{m.name}')}"""
+                for m in metadata)]
+
+        from_list = [
+            sqlite_quote_id(metadata[0].table_name),
+            *(f"full outer join {sqlite_quote_id(m.table_name)} on {sqlite_quote_id(m.table_name, m.id_column)} in ({', '.join(sqlite_quote_id(m.table_name, m.id_column) for m in reversed(preceding))})"
+                for m, preceding in [(m, metadata[:i]) for i, m in enumerate(metadata[1:], 1)])]
+
+        # Take some small pains to make the query readable since it makes
+        # debugging and development easier.  Note that backslashes aren't
+        # allowed inside f-string expressions, hence the *newline* variable.
+        newline = '\n'
+        query = dedent(f"""\
+            select
+                {(',' + newline + '                ').join(select_list)}
+            from
+                {(newline + '                ').join(from_list)}
+            ;
+            """)
+
+
+        # Write merged metadata as export from SQLite db.
+        #
+        # Assume TSV like nearly all other extant --output-metadata options.
+        print_info(f"Merging metadata and writing to {args.output_metadata!r}…")
+        sqlite3(db_path,
+            f'.mode tabs',
+            f'.headers on',
+            f'.once {sqlite_quote_dot(f"|{augur} write-file {shquote(args.output_metadata)}")}',
+            query)
```

Comment:
Did you consider the effects of TSV → sqlite → TSV conversion? In some quick testing I noticed that values with surrounding double quotes get stripped of the quotes. I'm curious if there are any other effects. I assume they would also be edge case-y and not something to worry about in practice, but still it would be good to know about them.

I had thought about this in 982bc1c93861d9e6e7b33d9a68e75b540fbc2938 when `augur filter` used TSV → pandas → TSV, but even with that I noted that nothing blocking has been brought to our attention.

Reply:
Good question! I did consider this, but didn't document the decision or pose the question to the group, and I _meant_ to do at least one if not both of those.

SQLite's `.mode tabs` is "standard" TSV, where tabs and newlines are simply not allowed in field values. If they appear, they break the output. There is no standard quoting or escaping mechanism.

SQLite's `.mode csv` with `.separator "\t"` is RFC-compliant CSV with tabs as the delimiter, where the standard quoting mechanism can be used to include tabs or newlines in field values.

It _seemed_ to me like Augur used/preferred the former, but maybe we actually prefer the latter? The latter seems better, frankly, and it's what we use on _import_ anyway. I should probably switch the output to use it as well too.

Reply:
> TSV → pandas → TSV

To me, the concern there is less the TSV format and more Pandas' deserialization and serialization of individual field values.

Reply:
> It _seemed_ to me like Augur used/preferred the former, but maybe we actually prefer the latter?

Seeing as we use Python's csv module to produce TSV, I think I got this wrong, yeah?

Reply:
Thanks for the details! +1 for CSV with tab delimiter to maintain consistency with existing use of csv module to produce TSV (as you've noted).

> > TSV → pandas → TSV
> 
> To me, the concern there is less the TSV format and more Pandas' deserialization and serialization of individual field values.

Agreed. This made me check if there is anything similar with SQLite. I think the answer is no under the following assumptions:

1. The type affinity is set to `TEXT` for all columns of an imported table.
    - I can't find a source for this apart from answers to this [StackOverflow question](https://stackoverflow.com/q/20240315), but some quick local testing backs this claim.
2. Any difference would only be observed if the output changes representation of a value based on its type affinity.
    - Example: a value stored as `"1"` with `REAL` affinity would be `1.0` in output.
3. A type affinity of `TEXT` does not change representation of values from any datatype.
    - I can't a source for this. The following line from the [Datatypes](https://www.sqlite.org/datatype3.html) doc raises concerns:
        > If numerical data is inserted into a column with TEXT affinity it is converted into text form before being stored.

        but some quick local testing with values such as `NULL`, `0`, `0.00`, `1.23e2`, `1.23000000e2`, `1.234567890123456789012345678901234567890` backs this claim.

Reply:
I [switched output](https://tsibley.net/paste/2024-08-01Rx9RoxrI-.html) to be CSV-like TSV. Quotes may be added or removed compared to the input, but parsed values should be equivalent. Most existing TSV produced by Augur uses the csv module's [`QUOTE_MINIMAL` quoting mode](https://docs.python.org/3/library/csv.html#csv.QUOTE_MINIMAL). The exception is `augur curate` commands which produce TSV with [`QUOTE_NONE` mode](https://docs.python.org/3/library/csv.html#csv.QUOTE_NONE), meaning it will error if field values contain a separator (tab) or newline. I've [commented about that](https://github.com/nextstrain/augur/issues/1312#issuecomment-2263713744) on the original PR.

Reply:
Updating `augur curate` to produce CSV-like TSVs in https://github.com/nextstrain/augur/pull/1565. 
Probably worth noting that this reminds me of the discussion on [Slack](https://bedfordlab.slack.com/archives/C01LCTT7JNN/p1713392993380609) of the dislike of CSV-like TSVs from `csvtk`.

Reply:
> This made me check if there is anything similar with SQLite. I think the answer is no under the following assumptions:

Thanks for double checking this. Good to have more 👀 and :brain: on it.

> The type affinity is set to `TEXT` for all columns of an imported table.
> 
>     * I can't find a source for this apart from answers to this [StackOverflow question](https://stackoverflow.com/q/20240315), but some quick local testing backs this claim.

I read thru the source of the `.import` routine a few weeks back and can confirm it treats everything as `text`: columns in the auto-created table are assigned `text` affinity and values are inserted as a bound parameter of type `text`.

Reply:
Tracking final decision about IANA TSVs (`.mode tabs`) vs. CSV-like TSVs (`.mode csv; .separator "\t"`) in https://github.com/nextstrain/augur/issues/1566.